lokimq/bt_serialize.h

```diff
@@ -909,5 +922,192 @@ class bt_dict_consumer : private bt_list_consumer {
     bt_dict_consumer consume_dict_consumer() { return consume_dict_data(); }
 };
 
+/// Currently this is not idiot-proof.  That is to say, the compiler won't catch everything and
+/// you technically can create invalid BEncoded strings.  Basically you can add list elements
+/// to a dict and dict elements to a list, and obviously that can end badly.  This may be fixed
+/// at a later time.
+struct fixed_buffer_producer
+{
+    char* begin;
+    char* end;
+
+    char* original_begin;
+
+    fixed_buffer_producer(char* begin, char* end) : begin(begin), end(end), original_begin(begin){
+        if (not (end > begin))
+            throw bt_serialize_length_error("Fixed buffer for serialization end pointer < begin pointer.");
+    }
+
+    // For now, used just for strings.  If we end up serializing large numbers as integers this
+    // will be used for that as well.  As this is designed for strings, will optimize for small
+    // integers and won't work for values > 99999.
+    size_t num_digits(size_t value)
+    {
+        size_t digits = 1;
+
+        if (value > 10) digits++;
+        if (value > 100) digits++;
+        if (value > 1000) digits++;
+        if (value > 10000) digits++;
```

Comment:
These should all be >= instead of > (10 has 2 digits, 100 has 3, etc.).  But also it would probably be faster to use `else if`s.  I'm also not a fan of the "won't work for values > 99999" bit.  This should be nice and fast for the common case, but won't break for a big value:
```C++
size_t digits;
if (value < 10) digits = 1;
else if (value < 100) digits = 2;
else if (value < 1000) digits = 3;
else if (value < 10000) digits = 4;
else {
    digits = 5;
    value /= 100000;
    while (value >= 10) { digits++; value /= 10; }
}
```

Reply:
why not just
```C++
while(value >= 10) { digits++; value /= 10; }
```

Reply:
makes sense.